src/cross_section.F90

```diff
@@ -49,89 +48,83 @@ subroutine calculate_xs(p)
     ! Exit subroutine if material is void
     if (p % material == MATERIAL_VOID) return
 
-    mat => materials(p % material)
-
-    ! Find energy index on energy grid
-    if (grid_method == GRID_MAT_UNION) then
-      i_grid = find_energy_index(mat, p % E)
-    else if (grid_method == GRID_LOGARITHM) then
+    associate (mat => materials(p % material))
+      ! Find energy index on energy grid
       i_grid = int(log(p % E/energy_min_neutron)/log_spacing)
-    end if
 
-    ! Determine if this material has S(a,b) tables
-    check_sab = (mat % n_sab > 0)
+      ! Determine if this material has S(a,b) tables
+      check_sab = (mat % n_sab > 0)
 
-    ! Initialize position in i_sab_nuclides
-    j = 1
+      ! Initialize position in i_sab_nuclides
+      j = 1
 
-    ! Add contribution from each nuclide in material
-    do i = 1, mat % n_nuclides
-      ! ========================================================================
-      ! CHECK FOR S(A,B) TABLE
+      ! Add contribution from each nuclide in material
+      do i = 1, mat % n_nuclides
+        ! ========================================================================
+        ! CHECK FOR S(A,B) TABLE
 
-      i_sab = 0
+        i_sab = 0
 
-      ! Check if this nuclide matches one of the S(a,b) tables specified -- this
-      ! relies on i_sab_nuclides being in sorted order
-      if (check_sab) then
-        if (i == mat % i_sab_nuclides(j)) then
-          ! Get index in sab_tables
-          i_sab = mat % i_sab_tables(j)
+        ! Check if this nuclide matches one of the S(a,b) tables specified -- this
+        ! relies on i_sab_nuclides being in sorted order
+        if (check_sab) then
+          if (i == mat % i_sab_nuclides(j)) then
+            ! Get index in sab_tables
+            i_sab = mat % i_sab_tables(j)
 
-          ! If particle energy is greater than the highest energy for the S(a,b)
-          ! table, don't use the S(a,b) table
-          if (p % E > sab_tables(i_sab) % threshold_inelastic) i_sab = 0
+            ! If particle energy is greater than the highest energy for the S(a,b)
+            ! table, don't use the S(a,b) table
+            if (p % E > sab_tables(i_sab) % data(1) % threshold_inelastic) i_sab = 0
```

Comment:
The `threshold_inelastic` value doesn't change with temperature (in fact, we could consider moving it in the HDF5 format to reflect that, but I'm kind of holding off because there are much larger changes I want to make to S(a,b) data organization).